granian/server.py

```diff
@@ -461,9 +465,50 @@ def setup_signals(self):
         if sys.platform != 'win32':
             signal.signal(signal.SIGHUP, self.signal_handler_reload)
 
-    def startup(self, spawn_target, target_loader):
-        logger.info(f'Starting granian (main PID: {os.getpid()})')
+    def _write_pid(self):
+        with self.pid.open('w+') as pid_file:
+            pid_file.write(f'{self.main_pid}\n')
+
+    def _write_pidfile(self):
+        if not self.pid:
+            return
+
+        if self.pid.exists():
+            with self.pid.open('r') as pid_file:
+                try:
+                    old_pid = int(pid_file.read())
+                except ValueError:
+                    self._write_pid()
+                    return
+            try:
+                os.kill(old_pid, 0)
+            except OSError as e:
+                if e.args[0] == errno.ESRCH:
+                    self._write_pid()
+                    return
+                else:
+                    raise e
+            if old_pid == self.main_pid:
+                return
+            msg = "Already running on PID %s (or pid file '%s' is stale)"
+            logger.error(msg % (old_pid, self.pid))
+            sys.exit(1)
```

Comment:
This should probably be a custom `Exception` that we catch in `cli.py` and exit there, in the same way we do for `ConfigurationError`.

Reply:
Addressed via https://github.com/emmett-framework/granian/compare/405ff32e5fbe32a1ac6db956e6d9df2c71913be8..ea4efa7dad1e4b1fbe09d2dfd822936be9d2095d

Let me know if the name of the custom `Exception` need to be changed.